perf: 백엔드 핫 리로딩 성능 벤치마크

scarf005 opened the pull request:
### 개요

> **Note** 벤치마크 툴링 및 기록 보관용 PR입니다. close해도 무방합니다. issue나 discussion으로 옮기는 편이 더 적절할 수도 있을 것 같습니다.

리팩터링을 하며 수정시마다 주기적으로 서버를 재시작해야 하는데, 기존 nodemon + ts-node 방식은 재시작 시간 (3.5초+) 부분에서 부족함을 느꼈습니다. 이에 다양한 파일 감시 솔루션을 찾아보았고, 벤치마크 기록 리포트를 작성했습니다.


사용한 방법은 `nodemon`, `tsx`, `ts-node-dev`, `swc`, `swc-node`, `node --watch` (wip) 등입니다.

`deno/` 경로에 테스트용 툴링이 들어 있으며, `autotouch.sh`로 100회씩 파일 수정을 하여 걸린 시간을 통계내 
[report.md](https://github.com/jiphyeonjeon-42/backend/blob/268907c5b66cec90fc0828ba99edf04cb601c1fb/deno/report.md)에 렌더링했습니다.

## Entity에서 ESM 순환 참조 문제

swc 사용시 ESM 스펙을 따르게 되는데, typeorm에서 엔티티간 relation (one-to-many 등) 타입을 파싱하는 과정에서, esm에서 허용하지 않는 순환 참조가 발생해 `Relation` 타입 래퍼를 추가해 해결했습니다.

https://github.com/typeorm/typeorm/issues/4526#issuecomment-1076462998

<details><summary>Details</summary>

```
 %  ~/r/l/backend   develop *$  backend  pnpm dev

> 42-jiphyeonjeon-backend-3rd@1.0.0 dev /home/scarf/repo/lib42/backend/backend
> nodemon --watch "src/**/*.ts" --exec "ts-node" --swc src/server.ts

[nodemon] 2.0.15
[nodemon] to restart at any time, enter `rs`
[nodemon] watching path(s): src/**/*.ts
[nodemon] watching extensions: ts,json
[nodemon] starting `ts-node --swc src/server.ts`
/home/scarf/repo/lib42/backend/backend/src
2023-06-27 19:51:43:5143 error: uncaughtException: Cannot access '_default' before initialization
ReferenceError: Cannot access '_default' before initialization
    at Object.get [as default] (/home/scarf/repo/lib42/backend/backend/src/entity/entities/User.ts:8:9)
    at Object.<anonymous> (/home/scarf/repo/lib42/backend/backend/src/entity/entities/Likes.ts:34:9)
    at Module._compile (node:internal/modules/cjs/loader:1255:14)
    at Module.m._compile (/home/scarf/repo/lib42/backend/backend/node_modules/.pnpm/ts-node@10.9.1_@swc+core@1.3.66_@types+node@17.0.15_typescript@4.5.5/node_modules/ts-node/src/index.ts:1618:23)
    at Module._extensions..js (node:internal/modules/cjs/loader:1309:10)
    at Object.require.extensions.<computed> [as .ts] (/home/scarf/repo/lib42/backend/backend/node_modules/.pnpm/ts-node@10.9.1_@swc+core@1.3.66_@types+node@17.0.15_typescript@4.5.5/node_modules/ts-node/src/index.ts:1621:12)
    at Module.load (node:internal/modules/cjs/loader:1113:32)
    at Function.Module._load (node:internal/modules/cjs/loader:960:12)
    at Module.require (node:internal/modules/cjs/loader:1137:19)
    at require (node:internal/modules/helpers:121:18)
 Error Stack: ReferenceError: Cannot access '_default' before initialization
    at Object.get [as default] (/home/scarf/repo/lib42/backend/backend/src/entity/entities/User.ts:8:9)
    at Object.<anonymous> (/home/scarf/repo/lib42/backend/backend/src/entity/entities/Likes.ts:34:9)
    at Module._compile (node:internal/modules/cjs/loader:1255:14)
    at Module.m._compile (/home/scarf/repo/lib42/backend/backend/node_modules/.pnpm/ts-node@10.9.1_@swc+core@1.3.66_@types+node@17.0.15_typescript@4.5.5/node_modules/ts-node/src/index.ts:1618:23)
    at Module._extensions..js (node:internal/modules/cjs/loader:1309:10)
    at Object.require.extensions.<computed> [as .ts] (/home/scarf/repo/lib42/backend/backend/node_modules/.pnpm/ts-node@10.9.1_@swc+core@1.3.66_@types+node@17.0.15_typescript@4.5.5/node_modules/ts-node/src/index.ts:1621:12)
    at Module.load (node:internal/modules/cjs/loader:1113:32)
    at Function.Module._load (node:internal/modules/cjs/loader:960:12)
    at Module.require (node:internal/modules/cjs/loader:1137:19)
    at require (node:internal/modules/helpers:121:18)
^C ELIFECYCLE  Command failed.

```

</details>

scarf005 committed:
fix: Entity에서 ESM 모듈 순환 참조 오류 해결

엔티티 간 관계 필드를 Relation 타입으로 래핑
손으로 일일이 하는거라 힘들었읍니다...

scarf005 committed:
chore: 시작 시간을 제하고 모든 로깅 제외

scarf005 committed:
chore: 재시작 시 `watcher:` 형식으로 로깅하도록 통일

scarf005 committed:
build: 리로딩 옵션 추가

scarf005 committed:
feat: 벤치마킹용 툴

scarf005 committed:
docs: 테스트 결과 리포트

편차가 큰 초기값이나 지나치게 평균에서 벗어난 값은 제거했습니다.

jimin52 commented:
`Entity에서 ESM 모듈 순환 참조 오류 해결` 커밋부분을 수요일에 설명 부탁드려도 될까요?
어떤 방식으로 속도 개선이 되었는지에 대해 공부하신 내용 공유 부탁드립니다 🙏

scarf005 committed:
fix: graceful shutdown 구현

nodemon등으로 빠르게 재시작시 기존 포트가 사용중이어서 오류가 발생하는 경우가 있습니다.
종료 및 재시작 시그널을 받을 시 기존 연결을 끊고 종료하도록 수정하였습니다.

scarf005 committed:
fix: 올바른 watcher-server 쌍만 검사하도록 변경

scarf005 committed:
docs: 리포트 업데이트

scarf005 committed:
perf: `node --watch` 기록 추가

scarf005 commented:
성능이 더 뛰어난 #537 이 있어 닫습니다.
